Linearizability checker memory reduction

henningandersen opened the pull request:
The cache used in linearizability checker now uses approximately 6x less
memory by changing the cache from a set of (bits, state) tuples into a
map from bits -> { state }.

Each combination of states is kept once only, building on the
assumption that the number of state permutations is small compared to
the number of bits permutations. For those histories that are difficult
to check we will have many bits combinations that use the same state
permutations.

We end up now using approximately 15 bytes per entry compared to 101
bytes before, ie. a 6x improvement, allowing us to linearizability check
significantly longer histories.

Re-enabled linearizability checker in CoordinatorTests, hoping above
ensures we no longer un out of memory.

Resolves #39437

henningandersen committed:
Linearizability checker memory reduction

The cache used in linearizability checker now uses approximately 6x less
memory by changing the cache from a set of (bits, state) tuples into a
map from bits -> { state }.

Each combination of states is kept once only, building on the
assumption that the number of state permutations is small compared to
the number of bits permutations. For those histories that are difficult
to check we will have many bits combinations that use the same state
permutations.

We end up now using approximately 15 bytes per entry compared to 101
bytes before, ie. a 6x improvement, allowing us to linearizability check
significantly longer histories.

Re-enabled linearizability checker in CoordinatorTests, hoping above
ensures we no longer un out of memory.

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch commented on server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java:
```diff
@@ -373,4 +374,81 @@ void unlift() {
+                }
+                objects = add(objects, state);
+            }
+            smallMap.put(bits, objects);
```
we can avoid an extra lookup here in case `smallMap.get(bits) != null` by using the `indexOf`, `indexGet` and `indexReplace` methods on the `LongObjectHashMap`.

henningandersen replied:
Fixed in 7853ddfcd5b675f0cc8c85a308b96f0588c381c7

ywelsch commented on server/src/test/java/org/elasticsearch/cluster/coordination/LinearizabilityChecker.java:
```diff
@@ -373,4 +374,81 @@ void unlift() {
+            return largeMap.computeIfAbsent(state, k -> new HashSet<>()).add(bitSet);
+        }
+
+        private Set<Object> add(Set<Object> objects, Object state) {
```
let's inline this method, it's only used once. Also this allows to move the call to `uniquePermutation` to the end of `addSmall` and allows inlining `uniquePermutation` as well.

henningandersen replied:
Fixed in 7853ddfcd5b675f0cc8c85a308b96f0588c381c7

ywelsch reviewed (approved):
Looks great. I've left two minor comments

henningandersen committed:
Linearizability checker memory reduction

Linearizability cache now use indexOf/indexGet/indexReplace/indexInsert
to optimize performance a bit.
